tests/conftest.py

```diff
@@ -450,21 +451,30 @@ def aws_credentials(monkeypatch):
 
 
 @pytest.fixture(scope="function")
-def mock_s3_resource(aws_credentials):
+def mock_s3_resource(aws_credentials, mock_s3_client):
     if moto_server_address is None:
         with moto.mock_aws():
             yield boto3.resource("s3", region_name=TEST_AWS_REGION_NAME)
     else:
         yield boto3.resource("s3", region_name=TEST_AWS_REGION_NAME, endpoint_url=moto_server_address)
 
 @pytest.fixture(scope="function")
-def mock_s3_client(aws_credentials):
+def get_mock_s3_client(aws_credentials):
     if moto_server_address is None:
         with moto.mock_aws():
             yield boto3.client("s3", region_name=TEST_AWS_REGION_NAME)
     else:
         yield boto3.client("s3", region_name=TEST_AWS_REGION_NAME, endpoint_url=moto_server_address)
 
+
+@pytest.fixture(scope="function")
+def mock_s3_client(get_mock_s3_client, monkeypatch):
+    def _get_client(*args, **kwargs):
+        return get_mock_s3_client
+
+    monkeypatch.setattr(S3ClientBuilder, "from_region", _get_client)
```

Comment:
is this monkeypathing actually necessary? Doesn't the moto library support this out of the box?

Reply:
For a lot of cases that is indeed true but we have test cases where the moto runs as the 'server' process but then a client mock should not follow ordinary moto mock logic but rather be a client configured to go to the endpoint of the moto server.